ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -168,31 +191,35 @@ public void run() {
   }
 
   static class RetryEvent {
-    private final KsqlEngine ksqlEngine;
     private final QueryId queryId;
     private final Ticker ticker;
+    private final long statusRunningThresholdMs;
 
     private int numRetries = 0;
     private long waitingTimeMs;
     private long expiryTimeMs;
     private long retryBackoffMaxMs;
     private long baseWaitingTimeMs;
+    private long healthyTimeMs;
 
     RetryEvent(
-        final KsqlEngine ksqlEngine,
         final QueryId queryId,
         final long baseWaitingTimeMs,
         final long retryBackoffMaxMs,
+        final long statusRunningThresholdMs,
         final Ticker ticker
     ) {
-      this.ksqlEngine = ksqlEngine;
       this.queryId = queryId;
       this.ticker = ticker;
+      this.statusRunningThresholdMs = statusRunningThresholdMs;
+
+      final long now = ticker.read();
 
       this.baseWaitingTimeMs = baseWaitingTimeMs;
       this.waitingTimeMs = baseWaitingTimeMs;
-      this.expiryTimeMs = ticker.read() + baseWaitingTimeMs;
       this.retryBackoffMaxMs = retryBackoffMaxMs;
+      this.expiryTimeMs = now + baseWaitingTimeMs;
+      this.healthyTimeMs = now + statusRunningThresholdMs;
```

Comment:
tracking this way is OK, but I think it would be better if we exposed how long the query was actually running from the query metadata. That way, we can know that what we're measuring is exactly query uptime, and not just time since the last attempted restart.

We can implement this using a `KafkaStreams.StateListener` to start the timer when transitioning to `RUNNING` for the first time.

Reply:
Done. I added this change in the last commit. The `QueryStateListener` tracks the up time when the state transitions to RUNNING or REBALANCING. This up time is obtained in the `QueryMonitor` through the `QueryMetadata.uptime()` call. Zero uptime is returned when the query is not running or rebalancing.